File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/LocksReleasedAllPaths.Monitor.TryEnter.cs
using System;
using System.Diagnostics;
using System.Threading;

namespace Monitor_TryEnter
{
    class Program
    {
        private object obj = new object();
        private bool condition;

        public void TryEnter_ExitedInElse()
        {
            if (Monitor.TryEnter(obj)) // Compliant, never exited in this method. It's a programming error not covered by this rule.
            {
            }
            else
            {
                Monitor.Exit(obj);
            }
        }

        public void TryEnter_ExitedInIf()
        {
            if (Monitor.TryEnter(obj)) // Noncompliant
            {
                if (condition)
                {
                    Monitor.Exit(obj);
                }
            }
        }

        public void TryEnter_Compliant()
        {
            if (Monitor.TryEnter(obj))
            {
                Monitor.Exit(obj);
            }
            else
            {
            }
        }

        public void TryEnter_WithInt()
        {
            if (Monitor.TryEnter(obj, 42)) // Noncompliant
            {
                if (condition)
                {
                    Monitor.Exit(obj);
                }
            }
        }

        public void TryEnter_WithRefIsAcquired(bool condition)
        {
            bool isAcquired = false;
            Monitor.TryEnter(obj, 42, ref isAcquired); // Noncompliant
            if (condition)
            {
                Monitor.Exit(obj);
            }
        }

        public void TryEnter_WithTimeSpan()
        {
            if (Monitor.TryEnter(obj, new TimeSpan(42))) // Noncompliant
            {
                if (condition)
                {
                    Monitor.Exit(obj);
                }
            }
        }

        public void Method6(bool condition)
        {
            bool isAcquired = false;
            Monitor.TryEnter(obj, new TimeSpan(42), ref isAcquired); // Noncompliant
            if (condition)
            {
                Monitor.Exit(obj);
            }
        }

        public void TryEnter_WithVariable_Compliant()
        {
            bool isAcquired = Monitor.TryEnter(obj, 42); // Compliant, never exited in this method

            if (isAcquired)
            {
            }
            else
            {
                Monitor.Exit(obj);
            }
        }

        public void TryEnter_WithVariable_Noncompliant()
        {
            bool isAcquired = Monitor.TryEnter(obj, 42); // Noncompliant

            if (isAcquired)
            {
                if (condition)
                {
                    Monitor.Exit(obj);
                }
            }
        }

        public void Method8()
        {
            bool isAcquired = Monitor.TryEnter(obj, 42);

            if (isAcquired)
            {
                Monitor.Exit(obj);
            }
        }

        public void Method9()
        {
            Monitor.TryEnter(obj); // Compliant
            Monitor.Exit(obj);
        }

        public void Method10()
        {
            Monitor.TryEnter(obj); // Noncompliant {{Unlock this lock along all executions paths of this method.}}
            if(condition)
                Monitor.Exit(obj);
        }

        public void Method11()
        {
            switch (Monitor.TryEnter(obj))
            {
                case true:
                    Monitor.Exit(obj);
                    break;
                default:
                    break;
            }
        }

        public void TryEnter_Switch_Compliant()
        {
            switch (Monitor.TryEnter(obj)) // Compliant, never exited in this method
            {
                case false:
                    Monitor.Exit(obj);
                    break;
                default:
                    break;
            }
        }

        public void TryEnter_Switch_Noncompliant()
        {
            switch (Monitor.TryEnter(obj)) // Noncompliant
            {
                case false:
                    break;
                default:
                    if (condition)
                    {
                        Monitor.Exit(obj);
                    }
                    break;
            }
        }

        public void Method13(bool condition)
        {
            bool isAcquired = false;
            Monitor.TryEnter(obj, 42, ref isAcquired);
            if (isAcquired)
            {
                Monitor.Exit(obj);
            }
        }

        public void Method14(bool condition)
        {
            bool isAcquired = false;
            Monitor.TryEnter(obj, ref isAcquired); // Noncompliant
            if (condition)
            {
                Monitor.Exit(obj);
            }
        }

        public void Method15(bool condition)
        {
            bool isAcquired = false;
            Monitor.TryEnter(obj, ref isAcquired);
            if (isAcquired)
            {
                Monitor.Exit(obj);
            }
        }

        public void TryEnterInsideIf_Finally()
        {
            if (Monitor.TryEnter(obj, 500)) // Compliant https://github.com/SonarSource/sonar-dotnet/issues/5415
            {
                try
                {
                }
                finally
                {
                    Monitor.Exit(obj);
                }
            }
        }

        public void TryEnterInsideIf_Finally_WithVar()
        {
            bool lockTaken = false;
            try
            {
                lockTaken = Monitor.TryEnter(obj); // Compliant https://github.com/SonarSource/sonar-dotnet/issues/5415
            }
            finally
            {
                if (lockTaken)
                    Monitor.Exit(obj);
            }
        }

        public void TryEnter_EarlyExit()
        {
            if (Monitor.TryEnter(obj) == false)
                return;

            try
            {
            }
            finally
            {
                Monitor.Exit(obj);
            }
        }
    }
}

File: analyzers/tests/SonarAnalyzer.Test/TestCases/SymbolicExecution/Roslyn/LocksReleasedAllPaths.Monitor.TryEnter.vb
Imports System.Threading

Namespace Monitor_TryEnter

    Class Program

        Private Condition As Boolean
        Private Obj As New Object()

        Public Sub Method1()
            If Monitor.TryEnter(Obj) AndAlso Condition Then ' Noncompliant
                Monitor.Exit(Obj)
            End If
        End Sub

        Public Sub Method2()
            If Monitor.TryEnter(Obj) Then
                Monitor.Exit(Obj)
            Else
            End If
        End Sub

        Public Sub Method3()
            If Monitor.TryEnter(Obj, 42) AndAlso Condition Then ' Noncompliant
                Monitor.Exit(Obj)
            End If
        End Sub

        Public Sub Method4()
            Dim IsAcquired As Boolean
            Monitor.TryEnter(Obj, 42, IsAcquired) ' Noncompliant
            If Condition Then Monitor.Exit(Obj)
        End Sub

        Public Sub Method5()
            If Monitor.TryEnter(Obj, New TimeSpan(42)) AndAlso Condition Then ' Noncompliant
                Monitor.Exit(Obj)
            End If
        End Sub

        Public Sub Method6()
            Dim IsAcquired As Boolean
            Monitor.TryEnter(Obj, New TimeSpan(42), IsAcquired) ' Noncompliant
            If Condition Then Monitor.Exit(Obj)
        End Sub

        Public Sub Method7()
            Dim IsAcquired As Boolean = Monitor.TryEnter(Obj, 42) ' Noncompliant
            If IsAcquired AndAlso Condition Then
                Monitor.Exit(Obj)
            End If
        End Sub

        Public Sub Method8()
            Dim IsAcquired As Boolean = Monitor.TryEnter(Obj, 42)
            If IsAcquired Then Monitor.Exit(Obj)
        End Sub

        Public Sub Method9()
            Monitor.TryEnter(Obj) ' Compliant
            Monitor.Exit(Obj)
        End Sub

        Public Sub Method10()
            Monitor.TryEnter(Obj) ' Noncompliant {{Unlock this lock along all executions paths of this method.}}
            If Condition Then Monitor.Exit(Obj)
        End Sub

        Public Sub Method12()
            Select Case Monitor.TryEnter(Obj) ' Noncompliant
                Case False
                Case Else
                    If Condition Then Monitor.Exit(Obj)
            End Select
        End Sub

        Public Sub Method13()
            Dim IsAcquired As Boolean
            Monitor.TryEnter(Obj, 42, IsAcquired)
            If IsAcquired Then Monitor.Exit(Obj)
        End Sub

        Public Sub Method14(Condition As Boolean)
            Dim IsAcquired As Boolean
            Monitor.TryEnter(Obj, IsAcquired) ' Noncompliant
            If Condition Then Monitor.Exit(Obj)
        End Sub

        Public Sub Method15()
            Dim IsAcquired As Boolean
            Monitor.TryEnter(Obj, IsAcquired)
            If IsAcquired Then Monitor.Exit(Obj)
        End Sub

    End Class

End Namespace
